File: mojo/android/javatests/validation_test_util.cc
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "mojo/android/javatests/validation_test_util.h"

#include "base/android/jni_android.h"
#include "base/android/jni_string.h"
#include "base/android/scoped_java_ref.h"
#include "base/test/test_support_android.h"
#include "jni/ValidationTestUtil_jni.h"
#include "mojo/public/cpp/bindings/tests/validation_test_input_parser.h"

namespace mojo {
namespace android {

bool RegisterValidationTestUtil(JNIEnv* env) {
  return RegisterNativesImpl(env);
}

jobject ParseData(JNIEnv* env, jclass jcaller, jstring data_as_string) {
  std::string input =
      base::android::ConvertJavaStringToUTF8(env, data_as_string);
  std::vector<uint8_t> data;
  size_t num_handles;
  std::string error_message;
  if (!test::ParseValidationTestInput(
          input, &data, &num_handles, &error_message)) {
    ScopedJavaLocalRef<jstring> j_error_message =
        base::android::ConvertUTF8ToJavaString(env, error_message);
    return Java_ValidationTestUtil_buildData(
               env, NULL, 0, j_error_message.obj()).Release();
  }
  void* data_ptr = &data[0];
  if (!data_ptr) {
    DCHECK(!data.size());
    data_ptr = &data;
  }
  jobject byte_buffer =
      env->NewDirectByteBuffer(data_ptr, data.size());
  return Java_ValidationTestUtil_buildData(env, byte_buffer, num_handles, NULL)
      .Release();
}

}  // namespace android
}  // namespace mojo

File: mojo/android/javatests/src/org/chromium/mojo/bindings/ValidationTestUtil.java
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.mojo.bindings;

import org.chromium.base.CalledByNative;
import org.chromium.base.JNINamespace;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

/**
 * Utility class for testing message validation. The file format used to describe a message is
 * described in The format is described in
 * mojo/public/cpp/bindings/tests/validation_test_input_parser.h
 */
@JNINamespace("mojo::android")
public class ValidationTestUtil {

    /**
     * Content of a '.data' file.
     */
    public static class Data {
        private final ByteBuffer mData;
        private final int mHandlesCount;
        private final String mErrorMessage;

        public ByteBuffer getData() {
            return mData;
        }

        public int getHandlesCount() {
            return mHandlesCount;
        }

        public String getErrorMessage() {
            return mErrorMessage;
        }

        private Data(ByteBuffer data, int handlesCount, String errorMessage) {
            this.mData = data;
            this.mHandlesCount = handlesCount;
            this.mErrorMessage = errorMessage;
        }
    }

    /**
     * Parse a '.data' file.
     */
    public static Data parseData(String dataAsString) {
        return nativeParseData(dataAsString);
    }

    private static native Data nativeParseData(String dataAsString);

    @CalledByNative
    private static Data buildData(ByteBuffer data, int handlesCount, String errorMessage) {
        ByteBuffer copiedData = null;
        if (data != null) {
            copiedData = ByteBuffer.allocateDirect(data.limit());
            copiedData.order(ByteOrder.LITTLE_ENDIAN);
            copiedData.put(data);
            copiedData.flip();
        }
        return new Data(copiedData, handlesCount, errorMessage);
    }
}
